Assets/_BForBoss/_Character/Scripts/PlayerWallRunBehaviour.cs

```diff
@@ -0,0 +1,232 @@
+
+        private readonly Vector3[] directions = new Vector3[]
+        {
+            Vector3.right, 
+            Vector3.right + Vector3.forward,
+            Vector3.forward,
+            Vector3.left + Vector3.forward,
+            Vector3.left
+        };
+
+        private bool _isWallRunning = false;
+        private Character _baseCharacter = null;
+        private FirstPersonCharacter _fpsCharacter = null;
+        private RaycastHit[] _hits;
+        private LayerMask _mask;
+        private Vector3 _lastWallRunPosition;
+        private Vector3 _lastWallRunNormal;
+        private float _baseMaxSpeed;
+        private Func<Vector2> _movementInput;
+
+        private float _currentJumpDuration = 0f;
+        private float _timeSinceWallAttach = 0f;
+        private float _timeSinceWallDetach = 0f;
+
+        private Transform ChildTransform
+        {
+            get
+            {
+                return _fpsCharacter != null ? _fpsCharacter.rootPivot : transform;
+            }
+        }
+
+        public void Initialize(Character baseCharacter, Func<Vector2> getMovementInput)
+        {
+            _baseCharacter = baseCharacter;
+            _fpsCharacter = baseCharacter as FirstPersonCharacter;
+            _movementInput = getMovementInput;
+            _mask = LayerMask.GetMask("ParkourWall");
+        }
+
+        public void Falling(Vector3 _)
+        {
+            if (!CanWallRun()) return;
+            _hits = new RaycastHit[directions.Length];
+            for (int i = 0; i < directions.Length; i++)
+            {
+                Vector3 dir = ChildTransform.TransformDirection(directions[i]); //convert directions to relative to player
+                Physics.Raycast(ChildTransform.position, dir, out _hits[i], _wallMaxDistance, _mask);
+                if (_hits[i].collider != null) Debug.DrawRay(ChildTransform.position, dir * _hits[i].distance, Color.green);
+                else Debug.DrawRay(ChildTransform.position, dir * _wallMaxDistance, Color.red);
+            }
+
+            if (GetSmallestRaycastHit(_hits, out RaycastHit hit))
+            {
+                WallRun(hit);
+            }
+            else if (_isWallRunning) StopWallRunning();
+        }
+
+        public void OnJumped(ref int _jumpCount)
+        {
+            _currentJumpDuration = 0f;
+            if (_isWallRunning)
+            {
+                StopWallRunning();
+                _jumpCount = 0;
+            }
+        }
+
+        public void OnLanded()
+        {
+            _currentJumpDuration = 0f;
+            if (_isWallRunning)
+            {
+                StopWallRunning();
+            }
+        }
+
+        public bool CanJump()
+        {
+            return _isWallRunning; // can always jump out of a wall run
+        }
+
+        public void OnWallRunning()
+        {
+            if (!_isWallRunning)
+            {
+                _timeSinceWallDetach += Time.deltaTime;
```

Comment:
more of a question, but should we be using `Time.FixedDeltaTime` instead since we are dealing with time in relation to movement operations?

Reply:
Hmm I think you might be right, I'll update those